src/examples/sudoku/sudoku-zkapp.ts

```diff
@@ -0,0 +1,158 @@
+
+  constructor(value: number[][]) {
+    super();
+    this.value = value.map((row) => row.map(Field));
+  }
+
+  hash() {
+    return Poseidon.hash(this.value.flat());
+  }
+}
+
+class SudokuZkapp extends SmartContract {
+  @state(Field) sudokuHash = State<Field>();
+  @state(Bool) isSolved = State<Bool>();
+
+  deploy(args: any) {
+    super.deploy(args);
+    this.self.update.permissions.setValue({
+      ...Permissions.default(),
+      editState: Permissions.proofOrSignature(),
+    });
+  }
+
+  @method submitSolution(sudokuInstance: Sudoku, solutionInstance: Sudoku) {
+    let sudoku = sudokuInstance.value;
+    let solution = solutionInstance.value;
+
+    // first, we check that the passed solution is a valid sudoku
+
+    // define helpers
+    let range9 = Array.from({ length: 9 }, (_, i) => i);
+    let oneTo9 = range9.map((i) => Field(i + 1));
+
+    function assertHas1To9(array: Field[]) {
+      oneTo9
+        .map((k) => range9.map((i) => array[i].equals(k)).reduce(Bool.or))
+        .reduce(Bool.and)
+        .assertEquals(true);
+    }
+
+    // check all rows
+    for (let i = 0; i < 9; i++) {
+      let row = solution[i];
+      assertHas1To9(row);
+    }
+    // check all columns
+    for (let j = 0; j < 9; j++) {
+      let column = solution.map((row) => row[j]);
+      assertHas1To9(column);
+    }
+    // check 3x3 squares
+    for (let k = 0; k < 9; k++) {
+      let [i0, j0] = divmod(k, 3);
+      let square = range9.map((m) => {
+        let [i1, j1] = divmod(m, 3);
+        return solution[3 * i0 + i1][3 * j0 + j1];
+      });
+      assertHas1To9(square);
+    }
+
+    // next, we check that the solution extends the initial sudoku
+    for (let i = 0; i < 9; i++) {
+      for (let j = 0; j < 9; j++) {
+        let cell = sudoku[i][j];
+        let solutionCell = solution[i][j];
+        // either the sudoku has nothing in it (indicated by a cell value of 0),
+        // or it is equal to the solution
+        Bool.or(cell.equals(0), cell.equals(solutionCell)).assertEquals(true);
+      }
+    }
+
+    // finally, we check that the sudoku is the one that was originally deployed
+    let sudokuHash = this.sudokuHash.get(); // get the hash from the blockchain
+    sudokuInstance.hash().assertEquals(sudokuHash);
+
+    // all checks passed => the sudoku is solved!
+    this.isSolved.set(Bool(true));
+  }
+}
+
+// setup
+const Local = Mina.LocalBlockchain();
+Mina.setActiveInstance(Local);
+const account1 = Local.testAccounts[0].privateKey;
+
+const zkappPrivkey = PrivateKey.random();
+let zkappAddress = zkappPrivkey.toPublicKey();
+
+async function deploy(sudoku: number[][]) {
+  let tx = await Mina.transaction(account1, () => {
+    const p = Party.createSigned(account1, { isSameAsFeePayer: true });
+    p.balance.subInPlace(Mina.accountCreationFee());
+    let zkapp = new SudokuZkapp(zkappAddress);
+    let sudokuInstance = new Sudoku(sudoku);
+    zkapp.deploy({ zkappKey: zkappPrivkey });
```

Comment:
Makes sense to rename `zkappKey` to `zkappPrivateKey` since that's how it's being used anyways :P

Reply:
I renamed `zkappPrivkey` -> `zkappKey` instead :D In typical crypto lingo, a "key" is a private key. A public key is an "address". So I don't think we have to be that verbose. (This also applies to `feePayerKey` vs `feePayerPrivateKey` @jasongitmail)